src/compiler/parser.ts

```diff
@@ -1811,7 +1811,7 @@ namespace ts {
         function parseEntityName(allowReservedWords: boolean, diagnosticMessage?: DiagnosticMessage): EntityName {
             let entity: EntityName = parseIdentifier(diagnosticMessage);
             while (parseOptional(SyntaxKind.DotToken)) {
-                const node = <QualifiedName>createNode(SyntaxKind.QualifiedName, entity.pos);
+                const node: QualifiedName = <QualifiedName>createNode(SyntaxKind.QualifiedName, entity.pos);  // !!!

```

Comment:
@weswigham good example! I think even if it the system is not always correct, it's better if it is conservative. So in your example, I could envision an acceptable solution where at the two property accesses, `x` is of type `Second | Third`. That way the implementation could get to a point where it is always either correct, or is too conservative because of a reassignment in a loop. And then loops could be fixed up later while people try out the feature.